### PR TITLE
Refactor Auth System: Migrate to Password-Based Auth with Email Verification, Route Guards, and Stability Fixes

### LINKED ISSUE #16: Replace Magic Link Auth with Email + Password Flow (Expo Go Compatibility)
## Summary

Our current use of Supabase magic link (passwordless) authentication is introducing unnecessary complexity, particularly in our ExpoGo-based environment for mobile development. While the approach aims to simplify login, it is instead creating friction and increasing engineering overhead.

This issue proposes replacing the magic link flow with a more predictable **email + password authentication**, combined with **one-time email verification during sign-up only**.

---

### 1. Expo Go / Deep Linking Limitations

* Magic link emails do not reliably redirect back into the Expo Go app.
* This breaks the authentication flow during development and testing.
* Deep linking configuration adds significant complexity.

### 2. Confusing User Experience

* Users must leave the app, open their email, and return via a link.
* This context switching feels slow and unintuitive.
* Many non-technical users:

  * Do not understand what a “magic link” is
  * Expect immediate login without checking email
* The flow can feel especially “hectic” on mobile devices.

### 3. Unpredictable Behavior

* Redirects vary across browsers, devices, and email clients
* Expired or reused links create additional confusion

### 4. Dependency on Email Delivery

* Login is blocked if the email is delayed or not received
* Users may retry multiple times, creating duplicate requests and frustration

### 5. Engineering Complexity

* Deep linking and redirect handling are difficult to implement and maintain
* Auth state management becomes more error-prone
* Significant time is spent debugging edge cases instead of building features

### 6. Harder Debugging & Support

* Failures are difficult to trace due to multiple points of failure:
  * Redirect URL issues
  * Expired or invalid tokens
* Reproducing issues locally (especially in Expo) is inconsistent

---

## Impact

* Slower development velocity due to auth-related debugging
* Higher support burden
* Overly complex implementation for a core feature

---

### Switch to Email + Password Authentication

* Standard and widely understood pattern
* Predictable and immediate login experience
* No reliance on email delivery during sign-in
* Simpler session handling and state management

### Add One-Time Email Verification (Post Sign-Up Only)

* Users verify their email once during account creation
* Maintains account authenticity without impacting login UX
* Eliminates repeated email dependency during sign-in

---

## Key Reasons for Change

1. **Improved User Experience**
   Removes unnecessary steps and confusion; aligns with user expectations.

2. **Better Compatibility with Expo Go**
   Avoids deep linking issues and broken redirect flows in Expo Go.

3. **Simpler Implementation**
   Easier to build, test, and maintain compared to magic link flows.

4. **Faster Debugging & Development**
   Fewer moving parts → quicker issue resolution and less engineering overhead.

5. **User Preference & Familiarity**
   Many users prefer the predictability of email + password authentication.

---

## Acceptance Criteria

* [ ] Replace magic link authentication with email + password login
* [ ] Implement sign-up flow with one-time email verification
* [ ] Remove magic link and redirect/deep linking logic
* [ ] Ensure stable session persistence across app restarts
* [ ] Update UI/UX to reflect the new authentication flow
* [ ] Test across Expo Go and production builds
* [ ] Validate improved sign-in success rate and reduced drop-offs

---

## Additional Notes

Magic links could optionally remain as a secondary login method in the future, but should not be the default due to the issues outlined above.

---

## Next Steps

* Align on authentication strategy
* Update Supabase auth configuration
* Implement new flow and migrate existing users if necessary

### PR DESCRIPTION
**Pull request by arjadas**
This PR introduces a complete overhaul of the authentication system, transitioning from passwordless to password-based authentication. It also adds email verification, improves route protection, and resolves several race conditions and error-handling issues across the auth flow.

**Key Changes**

**Authentication Overhaul**

- Migrated from passwordless to email + password authentication
- Rewrote and standardized all source-of-truth auth-related files
- Introduced a centralized auth service with session verification state
- Removed legacy/unused components (e.g., callback.tsx)

**Email Verification**

- Implemented email verification flow
- Updated user model to include verification status
- Added email verification middleware
- Allowed controlled access for unverified users on specific routes

**Routing & Access Control**

- Enforced auth and email verification guards at the routing level
- Improved middleware to handle nuanced access scenarios (e.g., partially onboarded users)

**Auth UI Updates**

- Migrated login flow to email/password with reset support
- Added PasswordStrengthMeter component
- Updated auth-related screens and JSON specifications

**Testing & Infrastructure**

Added tests for:
- Email verification middleware
- JWT token factory
- Application context behavior

**Bug Fixes**

- Fixed session timing race condition between:
  - Frontend auth state
  - Route guards
  - API interceptor behavior
- Improved auth error handling

**Documentation**

- Updated README with:
  - New auth flow
  - Tech stack details
- Added/cleaned up auth-related documentation

**Notable Refactors**

- Attempted full auth rewrite was reverted in favor of incremental stabilization
- Cleaned and simplified auth-related file structure (“lean auth”)

**Follow-Ups / Future Work**

- Consider adding multi-factor authentication (MFA)
- Expand test coverage for edge cases in auth flows
- Evaluate performance of auth guards under heavy routing

### CODE REVIEW

File: frontend/app/(auth)/reset-password.tsx [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+      setIsSubmitting(false);
+      return;
+    }
+
+    await authService.signOut();
+    setSuccessMessage("Your password has been updated. Please log in with your new password.");
+    setIsSubmitting(false);
+
+    router.replace("/(auth)/login");
+  };
```
[the remark this answers is not in the document]
Comment (arjadas) [reply frontend/app/(auth)/reset-password.tsx:106]: resolved. sent message through params.

File: README.md [hunk 2]
```diff
@@ -1,53 +1,274 @@
...
-3. Run the backend:
+Set backend configuration (environment variables or user-secrets):
 
-```powershell
-dotnet run --project TuitionIQ.csproj
+```bash
+ConnectionStrings__DefaultConnection=Host=<host>;Port=5432;Database=<db>;Username=<user>;Password=<password>;SSL Mode=Require;Trust Server Certificate=true
+Supabase__ProjectRef=<your-project-ref>
+Supabase__JwtSecret=<your-supabase-jwt-secret>
+App__AllowedOrigins__0=http://localhost:8081
```
[the remark this answers is not in the document]
Comment (arjadas) [reply README.md:124]: removed jwtsecret var

File: backend/src/TuitionIQ.Api/Middleware/UserActiveCheckMiddleware.cs [hunk 3]
```diff
@@ -37,45 +58,111 @@ public async Task InvokeAsync(HttpContext context, AppDbContext dbContext)
...
+  private static async Task WriteForbiddenAsync(HttpContext context, string code)
+  {
+    if (context.Response.HasStarted)
+    {
+      return;
+    }
+
+    context.Response.StatusCode = StatusCodes.Status403Forbidden;
+    await context.Response.WriteAsJsonAsync(new { code });
   }
```
[the remark this answers is not in the document]
Comment (arjadas) [reply backend/src/TuitionIQ.Api/Middleware/UserActiveCheckMiddleware.cs:167]: following best practice. changed to problem details format.
